### DISCUSSION TITLE
Hierarchical Queries?

### DISCUSSION THREAD

**Discussion by andoowhy** [bevyengine/bevy#1189] [Ideas]
I'd like to be able to make "hierarchical" queries, i.e. specify and iterate over an Entity's Components and its children's.

I wrote this for [my Unity3D ECS library](https://github.com/andoowhy/EgoCS/wiki/Systems-and-Constraints#parent-constraints), and in my experience it cut down on a lot of error-prone boilerplate, especially with collision events and React-styled UI composition.

**Comment by cart** [comment]
Sounds like we're on the same wavelength. I'm half way through a hierarchy rework that involves hierarchical queries :)

**Reply by Bendrien** [reply]
Awesome! I imagine this will allow something like `Query<(), With<Parent<With<SomeComponent>>>` or `Query<(), With<Child<With<SomeComponent>>>`?
Or maybe `Query<(Component1, Parent<(Component2, Component3)>)>`
What is the current state of the rework?

**Reply by Ratysz** [reply]
@Bendrien #1525

On a second thought, I'm not that sure this made it in.

**Reply by cart** [reply]
Yeah I postponed the hierarchy changes to work on Bevy ECS v2. My (very WIP/experimental) hierarchy changes:

* Made hierarchy parent/child operations transactional (the only way to add/remove children is through a World/Command builder or a HierarchyQuery)
* Added HierarchyQuery, which is actually just a type alias with extra methods implemented on top:
  ```rust
    pub trait Hierarchy {
        fn add_child(&mut self, parent: Entity, child: Entity) -> Result<(), AddChildError>;
    }

    type HierarchyQuery<'a, 'b, 'c, Filter = ()> =
        Query<'a, (Option<&'b mut Parent>, Option<&'c mut Children>), Filter>;

    impl<'a, 'b, 'c> Hierarchy for HierarchyQuery<'a, 'b, 'c> {
        fn add_child(&mut self, parent: Entity, child: Entity) -> Result<(), AddChildError> {
            // SAFE: unique access to children and parent. no overlap
            unsafe {
                let mut child_parent = self.get_component_unsafe::<Parent>(parent)?;
                if child_parent.0 == parent {
                    return Err(AddChildError::ParentAlreadyHasChild { child, parent });
                }

                let mut parent_children = self.get_component_unsafe::<Children>(parent)?;
                parent_children.0.push(child);
                child_parent.0 = parent;
            }

            Ok(())
        }
    }
    ```

HierarchyQuery sort of does what @Bendrien laid out, but it currently filters both parents and children. It probably makes sense to split the filter.

**Reply by cart** [reply]
(note that unsafe isn't need here. this should just use query.get_mut())

**Reply by lukors** [reply]
Do we have something like this yet? :)
